vehicle/mkvi/software/sensing/sensing.c

```diff
@@ -0,0 +1,69 @@
+/*
+Header:
+    Code for the Sensing Controller Board located in the nostril motherboard and
+the rear sensing box Author:
+    @author Richard Li
+    @date   03/25/2023
+*/
+
+/*----- Includes -----*/
+#include "sensing.h"
+#include "can_api.h"
+#include "log_uart.h"
+#include "utils/timer.h"
+#include <avr/interrupt.h>
+#include <avr/io.h>
+#include <stdio.h>
+#include <string.h>
+#include <util/delay.h>
+
+void timer1_isr(void) {
+    send_can = true;
+};
+
+int pcint0_callback(int counter_left) {
+    // Left wheel speed sensor is in pin PB2 -> PORTB
+    sensing.wsl = !gpio_get_pin(WHEEL_SPEED_LEFT);
+    if (sensing.wsl) {
+        counter_left += 1;
+    }
+}
+void pcint2_callback(int counter_right) {
+    // Right wheel speed sensor is in pin PB2 -> PORTB
+    sensing.wsr = !gpio_get_pin(WHEEL_SPEED_RIGHT);
+    if (sensing.wsr) {
+        counter_right += 1;
+    }
+}
+
+int main(void) {
+    can_init_sensing();
+    adc_init();
+    CAN_init(CAN_ENABLED);
+    gpio_set_mode(DEBUG_LED_0, OUTPUT);
+    gpio_set_mode(WHEEL_SPEED_LEFT, INPUT);
+    gpio_set_mode(WHEEL_SPEED_RIGHT, INPUT);
+
+    sei(); // Initializes interrupts
+    gpio_enable_interrupt(WHEEL_SPEED_LEFT);
+    gpio_enable_interrupt(WHEEL_SPEED_RIGHT);
+
+    timer_init(&timer1_cfg);
+
+    pcint0_callback();
+    pcint2_callback();
+    timer1_isr();
+    while (true) {
+        // Check if the timer is at 1 second (arbitrary)
+        // If timer is at 1 second
+        if (send_can) {
+            // Log how many times we saw metal
+            sensing.left_wheel_counter = counter_left;
```

Comment:
Also, something I just thought of that would reduce our memory footprint and might be a little easier - instead of incrementing separate counters, and then assigning those values into our CAN struct, we can just increment the struct fields in our GPIO interrupt. So they might look something like:

```
void pcint2_callback(int counter_right) {
    // Right wheel speed sensor is in pin PB2 -> PORTB
    sensing.wsr = !gpio_get_pin(WHEEL_SPEED_RIGHT);
    if (sensing.wsr) {
        sensing.right_wheel_counter += 1;
    }
}
```

and then in this block you can just send CAN and reset your counters to 0

Reply:
Additionally, those functions cannot take arguments because they are also used in the GPIO library where they are called without arguments.